File: Kotlin/src/main/kotlin/intermediate/boj_10830.kt
package intermediate

private const val DIV = 1000UL
private const val MAX = 5
private val Matrix = Array(MAX) { ulongArrayOf() }

private fun matmul(A: Array<ULongArray>, B: Array<ULongArray>, n: ULong): Array<ULongArray> {
    val size = n.toInt()
    val ans = Array(size) { ULongArray(size) { 0UL } }

    for (k in 0 until size)
        for (i in 0 until size)
            for (j in 0 until size)
                ans[k][i] += (A[k][j] * B[j][i]) % DIV

    return ans
}

private fun solve(A: Array<ULongArray>, n: ULong, b: ULong): Array<ULongArray> {
    if (b == 1UL) return A

    val tmp = solve(A, n, b / 2UL)

    if (b % 2UL == 1UL) {
        val tmp2 = matmul(tmp, tmp, n)
        return matmul(tmp2, A, n)
    }

    return matmul(tmp, tmp, n)
}

fun boj_10830() = with(System.`in`.bufferedReader()) {
    val (n, b) = readLine()
        .split(" ")
        .map { it.toULong() }

    val size = n.toInt()

    for (i in 0 until size) {
        val tmp = readLine()
            .split(" ")
            .map { it.toULong() }
            .toULongArray()

        Matrix[i] = tmp
    }

    val ans = solve(Matrix, n, b)

    for (i in 0 until size) {
        for (j in 0 until size)
            print("${ans[i][j] % DIV} ")
        println()
    }
}

File: Cpp/Intermediate/boj_10830.cpp
//
// Created by seungsu on 11/30/22.
//

// 재귀를 이용한 방법 -> 시간초과

#include <iostream>
#include <vector>

using namespace std;

typedef unsigned long long ull;
typedef vector<vector<ull>> Matrix;

constexpr int MAX = 5;
constexpr int DIV = 1000;

static Matrix mat(MAX, vector<ull>(MAX));

static Matrix operator*(const Matrix& A, const Matrix& B) {
    int size = A.size();
    Matrix tmp(size, vector<ull>(size));

    for (int k = 0; k < size; k++) {
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++)
                tmp[k][i] += (A[k][j] * B[j][i]) % DIV;
        }
    }

    return tmp;
}

static Matrix solve(const Matrix& A, ull times) {
    if (times == 1) return A;

    Matrix tmp = solve(A, times / 2);

    if (times & 1) {
        Matrix tmp2 = tmp * A;
        return tmp * tmp2;
    }

    return tmp * tmp;
}

void boj_10830() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    ull n, b;
    cin >> n >> b;

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++)
            cin >> mat[i][j];
    }


    Matrix&& ans = solve(mat, b);

    for (int i = 0; i < n; i++) {
         for (int j = 0; j < n; j++)
             cout << ans[i][j] % DIV << " ";
         cout << "\n";
    }
}
